File: web/src/features/notification/lib/review-deep-link.ts
// Review deep-link helper.
//
// Some notification types ("Mission terminée — laissez un avis",
// "review_received") must navigate to the messaging surface AND ask
// it to auto-open the review modal for the right proposal. The
// backend (dispute & proposal flows) ships the navigation hints in
// `notification.data`:
//
//   {
//     "proposal_id":     "<uuid>",
//     "conversation_id": "<uuid>",
//     "proposal_title":  "..."
//   }
//
// `buildReviewDeepLink` reads that payload and produces the canonical
// URL the messaging-page understands:
//
//   /messages?id=<convId>&openReview=1&reviewProposalId=<proposalId>
//
// Returns `null` when the notification kind is not review-related OR
// when the payload is missing/malformed — the caller then falls back
// to the legacy "mark as read only" behaviour.

import type { Notification, NotificationType } from "../types"

const REVIEW_KINDS: ReadonlySet<NotificationType> = new Set<NotificationType>([
  // Both completion flows (proposal + dispute) fire this kind with
  // the "leave a review" call-to-action body. Distinct from the
  // "review_received" kind which fires when the counterparty has
  // already submitted theirs.
  "proposal_completed",
  "review_received",
])

function readString(data: Record<string, unknown>, key: string): string | null {
  const value = data[key]
  return typeof value === "string" && value.length > 0 ? value : null
}

export function buildReviewDeepLink(notification: Notification): string | null {
  if (!REVIEW_KINDS.has(notification.type)) return null

  const data = (notification.data ?? {}) as Record<string, unknown>
  const conversationId = readString(data, "conversation_id")
  const proposalId = readString(data, "proposal_id")
  if (!conversationId || !proposalId) return null

  const params = new URLSearchParams({
    id: conversationId,
    openReview: "1",
    reviewProposalId: proposalId,
  })
  return `/messages?${params.toString()}`
}

File: mobile/lib/features/notification/domain/review_deep_link.dart
import 'entities/app_notification.dart';

/// Review deep-link helper for mobile in-app notification taps.
///
/// Mirrors the web `buildReviewDeepLink` contract: review-related
/// notification kinds (proposal_completed, review_received) ship a
/// proposal-flow payload (`proposal_id` + `conversation_id` +
/// `proposal_title`) and tapping them should open the conversation
/// route with the openReview flag so the chat screen auto-opens the
/// review sheet for the right proposal.
///
/// Returns `null` when the notification kind is not review-related OR
/// when the payload is missing/malformed — the caller (the
/// NotificationScreen tap handler) then falls back to plain
/// mark-as-read with no navigation.
const _reviewKinds = {
  'proposal_completed',
  'review_received',
};

class ReviewDeepLink {
  final String conversationId;
  final String proposalId;

  const ReviewDeepLink({
    required this.conversationId,
    required this.proposalId,
  });

  /// Builds the go_router-friendly path for the chat screen:
  /// `/chat/<convId>?openReview=1&reviewProposalId=<proposalId>`.
  String toRoute(String chatBasePath) {
    return '$chatBasePath/$conversationId'
        '?openReview=1&reviewProposalId=$proposalId';
  }
}

/// Reads the navigation hints from the notification payload. Mirrors
/// the web shape exactly so a notification produced by the backend
/// dispute / proposal flow yields a usable deep-link on both surfaces.
ReviewDeepLink? buildReviewDeepLink(AppNotification notification) {
  if (!_reviewKinds.contains(notification.type)) return null;

  final conversationId = _readString(notification.data, 'conversation_id');
  final proposalId = _readString(notification.data, 'proposal_id');
  if (conversationId == null || proposalId == null) return null;

  return ReviewDeepLink(
    conversationId: conversationId,
    proposalId: proposalId,
  );
}

String? _readString(Map<String, dynamic> data, String key) {
  final value = data[key];
  if (value is String && value.isNotEmpty) return value;
  return null;
}
